File: benchmarks/trisolve/trisolve.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <math.h>
#include <lapacke.h>

#define DEV_MODE 1 

void reset() {}

void printMatrix(int n, int m, double *matrix) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            printf("%.6lf ", matrix[i * m + j]);
        }
        printf("\n");
    }
}

void initialize(int N, double *L, double *x, double *b) {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            L[i * N + j] = (i + N - j + 1) * 2.0 / N;
        }
        x[i] = -999.0;
        b[i] = (double)i;
    }
}

void kernel(int N, double *L, double *x, double *b) {
    for (int i = 0; i < N; i++) {
        double dp = 0.0;
        for (int j = 0; j < i; j++) {
            dp += L[i * N + j] * x[j];
        }
        x[i] = (b[i] - dp) / L[i * N + i];
    }
}

bool is_correct(int N, double *L, double *x, double *b) {
    int pivots[N];
    int result;

    double L_linear[N * N];
    for (int j = 0; j < N; j++) {
        for (int i = 0; i < N; i++) {
            if (i > j) {
                L[i * N + j] = 0.0;
            }
            L_linear[i + j * N] = L[i * N + j];
        }
    }

    result = LAPACKE_dgesv(LAPACK_COL_MAJOR, N, 1, L_linear, N, pivots, b, N);

    if (result == 0) {
        for (int i = 0; i < N; i++) {
            x[i] = b[i]; // solution stored in the 'b' array
        }
        return true;
    } else {
        return false;
    }
}

int main() {
    int N = 4;
    double *L = malloc(sizeof(double) * N * N);
    double *x = malloc(sizeof(double) * N);
    double *b = malloc(sizeof(double) * N);

    initialize(N, L, x, b);

    #if DEV_MODE
    printf("Data\n");
    printMatrix(N, N, L);

    printf("x:\n");
    printMatrix(1, N, x);

    printf("b:\n");
    printMatrix(1, N, b);
    #endif

    kernel(N, L, x, b);
    
    #if DEV_MODE
    printf("Result (x):\n");
    printMatrix(1, N, x);
    #endif

    if (is_correct(N, L, x, b)) {
        printf("Test passed: The solution matches the expected result.\n");
    } else {
        printf("Test failed: The solution does not match the expected result.\n");
    }

    return 0;
}

File: benchmarks/trisolve/trisolve.jl
using LinearAlgebra
include("../../timing/dphpc_timing.jl")

DEV = true
TIME = true
DEBUG = false

eval_benchmarks = Dict(
    "S" => 2000,
    "M" => 5000,
    "L" => 14000,
    "paper" => 16000
)

dev_benchmarks = Dict(
    "S" => 4,
    "M" => 6
)

function main()
    if DEV benchmark_sizes = dev_benchmarks 
    else benchmark_sizes = eval_benchmarks end

    validate()

    for (preset, N) in benchmark_sizes
        L, x, b = initialize(N, eltype(Float64))

        if DEBUG print_all(L, x, b) end

        if TIME
            println("Benchmarking $preset")
            res = @dphpc_time nothing kernel(L, x, b) 
            println(res)
        else
            x = kernel(L, x, b)
            if DEBUG println(x) end
        end
    end
end


function kernel(L, x, b)
    N = length(x)
    for i in 1:N
        x[i] = (b[i] - dot(L[i, 1:i-1], x[1:i-1])) / L[i, i]
    end
    return x
end

function initialize(N, datatype=Float64)
    L = [((i + N - j + 1) * 2 / N) for i in 1:N, j in 1:N]

    x = fill(-999.0, N)
    b = [i - 1 for i in 1:N]

    return L, x, b
end

function validate()
    args = initialize(5)
    kernel(args...)
    if !is_correct(args...)
        throw("Validation failed")
    end
end

function is_correct(L, x, b)
    N = size(L, 1)
    for i in 1:N, j in 1:N
        if i < j
            L[i, j] = 0
        end
    end

    expected_solution = L\b
    println(expected_solution)

    # Compare the result with the expected solution
    if isapprox(x, expected_solution)
        println("Test passed: The solution matches the expected result.")
        display(x)
        return true
    else
        println("Test failed: The solution does not match the expected result.")
        return false
    end
end

function print_all(L, x, b)
    println("Data")
    println("Running on set N=$(size(L, 1))")
    println("L:")
    display(L)
    println("x:")
    display(x)
    println("b:")
    display(b)
end

function reset() end

main()
